source/dhtslib/vcf/reader.d

```diff
@@ -160,39 +162,43 @@ struct VCFReaderImpl(CoordSystem cs, bool isTabixFile)
     /// InputRange interface: iterate over all records
     @property bool empty()
     {
-        //     int bcf_read(htsFile *fp, const(bcf_hdr_t) *h, bcf1_t *v);

```

Comment:
Overall the code is very similar. Just in order to be follow phobos rules, the range `empty` method cannot move the range forward only `popFront` can. I have solved this in our other ranges just as I have here, moved the "iterator next" call to `popFront` and introduced an `initialized` variable that is set when you check if the range is empty. This allows your to copy the range and still work as intended and allow other ranges to copy it as well. The issue in the first place was that it would hinder the ability of say `map` or `filter` or other phobos range functions to correctly use the range, as they expect to call `empty` without moving the range forward. I had already fixed this issue with the `SAMReader` `RecordRange` as I ran into it several months ago.

Reply:
As long as the range is extensively tested and passes -- including things like reading from an already expended range, etc. then I am good with this